### ISSUE TITLE
Unable to retrieve historical performance metrics on vCenter v7.0

### ISSUE THREAD

**Issue by jokjur** [vmware/govmomi#2438]
I'm retrieving performance metrics for virtual machines using govmomi (v0.25.0)

Real-time metrics are successfully retrieved in vCenter versions 6.5, 6.7 and 7.0. However, historical metrics (i.e intervalId 7200 or 86400) are only retrieved for vCenter 6.5 and 6.7, whilst vCenter 7.0 fails with following error:

```
ServerFaultCode: This operation is restricted by the administrator - 'vpxd.stats.maxQueryMetrics'. Contact your system administrator.
```

Code example used for metrics retrieval:
```
	querySpec := types.PerfQuerySpec{
		StartTime: &start,
		MetricId:   []types.PerfMetricId{{Instance: ""}},
		IntervalId: 7200,
	}

	sample, err := performanceManager.SampleByName(ctx, querySpec,
		metrics, vms)
```

In the provided snippet, `metrics` is list of metrics to retrieve, that consists of about 15 metrics. All vCenters use default configurations, therefore default value 256 of `vpxd.stats.maxQueryMetrics` should not be reached. On top of that, in our use case manually changing maxQueryMetrics config parameter to unlimited would not be a acceptable solution. 

Furthermore, I've tested the same flow with VMWare Java SDK and it successfully retrieves same performance metrics for vCenter 7.0

Is this a bug or what might be the root cause for this problem?

**Comment by dougm** [comment]
I don't know if/where this is documented, so I looked at the vCenter source code.   From what I can see, maxQueryMetrics does not apply to the realtime metric interval, only historical intervals.  For historical, the count is `metrics x entities` (vms in this case).  And if an entity does not have any stats for the given interval, it does not count against the max. Which I think is possible for newer VM instances that don't haven't rollup data yet for the given interval.  How many vms are you trying to sample?

**Comment by jokjur** [comment]
> 
> 
> I don't know if/where this is documented, so I looked at the vCenter source code. From what I can see, maxQueryMetrics does not apply to the realtime metric interval, only historical intervals. For historical, the count is `metrics x entities` (vms in this case). And if an entity does not have any stats for the given interval, it does not count against the max. Which I think is possible for newer VM instances that don't haven't rollup data yet for the given interval. How many vms are you trying to sample?

Thank you for your answer. We're trying to sample metrics for around 20 vms. By information you provided, that maxQueryMetrics is counted as `metrics x entities` the error would make sense as it would higher than 256. However, the interesting thing is that code written with `VMware Web services API Java SDK` does not throw this error. I've tried reducing the amount of metrics retrieved and it seems govmomi then successfully pulls the metrics. After successfully retrieving smaller batch of metrics with govmomi I tried to include metrics that vms do not have any stats for the given interval (i.e vm is too new or that metric collection is disabled) and govmomi then again responded with `maxQueryMetrics` error. By the looks of things, I feel that there might be problem that the mentioned VMware Web Services Java sdk does not count metrics with no information for the entity against the max and govmomi does.
